sklearn/metrics/_classification.py

```diff
@@ -2649,3 +2649,547 @@ def brier_score_loss(y_true, y_prob, *, sample_weight=None, pos_label=None):
+    ... labels=['pig', 'dog', 'cat'])
+    (array([0., 0., 1.]), array([0.25, 0.5 , 0.25]),
+    array([0.75, 0.5 , 0.75]), array([1., 1., 0.]))
+
+    Notes
+    -----
+    When `true positive + false negative == 0`, TPR, FNR are undefined;
+    When `true negative + false positive == 0`, FPR, TNR are undefined.
+    In such cases, by default the metric will be set to 0, as will F-score,
+    and `UndefinedMetricWarning` will be raised. This behavior can be
+    modified with `zero_division`.
+    """
+    _check_zero_division(zero_division)
+
+    labels = _check_set_wise_labels(y_true, y_pred, average, labels, pos_label)
+
+    # Calculate tp_sum, fp_sum, tn_sum, fn_sum, pos_sum, neg_sum
+    samplewise = average == "samples"
+    MCM = multilabel_confusion_matrix(
+        y_true,
+        y_pred,
+        sample_weight=sample_weight,
+        labels=labels,
+        samplewise=samplewise,
+    )
+    tp_sum = MCM[:, 1, 1]
+    fp_sum = MCM[:, 0, 1]
+    tn_sum = MCM[:, 0, 0]
+    fn_sum = MCM[:, 1, 0]
+    pos_sum = tp_sum + fn_sum
+    neg_sum = tn_sum + fp_sum
+
+    if average == "micro":
+        tp_sum = np.array([tp_sum.sum()])
+        fp_sum = np.array([fp_sum.sum()])
+        tn_sum = np.array([tn_sum.sum()])
+        fn_sum = np.array([fn_sum.sum()])
+        pos_sum = np.array([pos_sum.sum()])
+        neg_sum = np.array([neg_sum.sum()])
+
+    # Divide, and on zero-division, set scores and/or warn according to
+    # zero_division:
+    tpr = _prf_divide(
+        tp_sum, pos_sum, "TPR", "positives", average, warn_for, zero_division
+    )
+    fpr = _prf_divide(
+        fp_sum, neg_sum, "FPR", "negatives", average, warn_for, zero_division
+    )
+    tnr = _prf_divide(
+        tn_sum, neg_sum, "TNR", "negatives", average, warn_for, zero_division
+    )
+    fnr = _prf_divide(
+        fn_sum, pos_sum, "FNR", "positives", average, warn_for, zero_division
+    )
+    # Average the results
+    if average == "weighted":
+        weights = pos_sum
+        if weights.sum() == 0:
+            zero_division_value = 0.0 if zero_division in ["warn", 0] else 1.0
+            # TPR and FNR is zero_division if there are no positive labels
+            # FPR and TNR is zero_division if there are no negative labels
+            return (
+                zero_division_value if pos_sum.sum() == 0 else 0,
+                zero_division_value if neg_sum.sum() == 0 else 0,
+                zero_division_value if neg_sum.sum() == 0 else 0,
+                zero_division_value if pos_sum.sum() == 0 else 0,
+            )
+
+    elif average == "samples":
+        weights = sample_weight
+    else:
+        weights = None
+
+    if average is not None:
+        assert average != "binary" or len(fpr) == 1
+        tpr = np.average(tpr, weights=weights)
+        fpr = np.average(fpr, weights=weights)
+        tnr = np.average(tnr, weights=weights)
+        fnr = np.average(fnr, weights=weights)
+    return tpr, fpr, tnr, fnr
```

Comment:
This comment is orthogonal to this PR and will not require any action.
I am wondering about the API here and the compatibility with multimetric.

I think that it would be nice to support multimetric in `GridSearchCV` with the minumum of code. I would expect the following most probably:

```python
scorer = make_scorer(tpr_fpr_tnr_fnr_score)
GridSearchCV(..., scoring=scorer)
```

It would only work if `make_scorer` was able to return a dict I think. I am wondering if we could have an API that add a `return_type` to the metric function returning multiple scalars such that we can decide to return an array (2D as in confusion matrix), a tuple (like here and other metrics) or a Python dict with the name and the metric associated.

In this manner, `make_scorer` can understand `dict` and output a dict that can directly be used by the `scoring` of the `GridSearchCV` without the need of defining a custom callable to simply rewrap the scalar into a `dict`.

@jnothman do you have any thought. I am almost sure that this aleady something that you might have discussed when creating the multimetric API